src/backend/distributed/executor/local_executor.c

```diff
@@ -175,6 +177,120 @@ ExecuteLocalTaskList(CitusScanState *scanState, List *taskList)
 }
 
 
+/*
+ * LocalTaskPlannedStmt creates the PlannedStmt for the input task.
+ *
+ * The function behaves differently when the task->queryString is avaliable
+ * or not. See the comments in the function for the details.
+ */
+static Query *
+LocalShardQuery(Task *task, ParamListInfo boundParams,
+				int *numParams, Oid **parameterTypes)
+{
+	/*
+	 * For performance reasons, the queryString is not generated for
+	 * local-fast path queries. Instead, we simply update the shard
+	 * references to the actual shard relations.
+	 *
+	 * Otherwise, like multi-shard queries going through the local execution,
+	 * we'd need to parse the queryString back to Query before passing it
+	 * to the planner,
+	 */
+	if (task->query != NULL)
+	{
+		Query *shardQuery = copyObject(task->query);
+
+		ReplaceShardReferencesWalker((Node *) shardQuery, task);
+
+#if PG_VERSION_NUM < 120000
+		AcquireRewriteLocks(shardQuery, true, false);
+#else
+
+		/*
+		 * We acquire the required locks for PG12+ inside
+		 * ReplaceShardReferencesWalker.
+		 */
+#endif
+		return shardQuery;
+	}
+
+	if (*numParams == -1)
+	{
+		*numParams = 0;
+
+		if (boundParams != NULL)
+		{
+			const char **parameterValues = NULL;     /* not used anywhere, so decleare here */
+
+			ExtractParametersForLocalExecution(boundParams, parameterTypes,
+											   &parameterValues);
+
+			*numParams = boundParams->numParams;
+		}
+	}
+
+	return ParseQueryString(TaskQueryString(task), *parameterTypes, *numParams);
+}
+
+
+/*
+ * ReplaceShardReferencesWalker update RTE_RELATIONs that are the distributed
+ * relations to RTE_RELATIONs that are the local shards. In PG12+ we acquire
+ * the required locks on the local shards when we replace them, for PG11 these
+ * locks should be taken manually.
+ */
+static bool
+ReplaceShardReferencesWalker(Node *node, Task *task)
+{
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	if (IsA(node, RangeTblEntry))
+	{
+		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) node;
+
+		/*
+		 * We should have converted all RTE_RELATIONs to
+		 * citus_extradata_container function calls
+		 */
+		Assert(rangeTableEntry->rtekind != RTE_RELATION);
+
+		if (rangeTableEntry->rtekind == RTE_FUNCTION)
+		{
+			UnsetRangeTblExtraData(rangeTableEntry);
+		}
+
+		/* RTE_FUNCTION can be changed to a RTE_RELATION after UnsetTblExtraData */
+		if (rangeTableEntry->rtekind == RTE_RELATION)
+		{
+#if PG_VERSION_NUM >= 120000
+			LockRelationOid(rangeTableEntry->relid, rangeTableEntry->rellockmode);
```

Comment:
What's the reason for this difference between versions?

Reply:
rellockmode was introduced in PG12. It's also the only version that actually complained about locks not being taken.


I think the rellockmode approach is much cheaper, since there's no extra tree walking and it's very easy to reason about.